### PR TITLE
Small script which checks, whether a mesh is correct.

### PR DESCRIPTION
**Pull request by sebwolf-de**
See readme for details.

### CODE REVIEW

File: check_correctness/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+```
+mkdir build && cd build
+cmake ..
+make
+```
+
+## Usage
+
+```
+mpirun -n 4 ./check_correctnes -m mesh.h5
```
Comment (Thomas-Ulrich) [inline check_correctness/README.md]: typo

File: check_correctness/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+## Description?
```
Comment (Thomas-Ulrich) [inline check_correctness/README.md]: remove "?"

File: check_correctness/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+1. Read `mesh.h5`
+2. For each element check that:
+  - if a face is a regular or a dynamic rupture face, the neighbor *has to exist*.
+  - if a face is a free-surface or an absorbing face, the neighbor *must not exist*.
+3. The program aborts, when the first face is found that breaks the rules above.
```
Comment (Thomas-Ulrich) [inline check_correctness/README.md]: The program aborts when the first face that breaks the above rules is found.?

File: check_correctness/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+4. If all elements are correct, the programm exits with a succes output.
```
Comment (Thomas-Ulrich) [inline check_correctness/README.md]: success

File: check_correctness/src/mesh.cpp [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#include "PUML/Downward.h"
+
+Mesh::Mesh(const std::string& fileName) {
+  puml.open((fileName + ":/connect").c_str(), (fileName + ":/geometry").c_str());
+  puml.addData((fileName + ":/group").c_str(), PUML::CELL);
+  puml.addData((fileName + ":/boundary").c_str(), PUML::CELL);
+  int numPartitionCells = puml.numOriginalCells();
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  // Keep all cells on the same rank
```
Comment (Thomas-Ulrich) [inline check_correctness/src/mesh.cpp]: so you read the full mesh on all ranks?

File: check_correctness/src/mesh.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  puml.partition(partition.data());
+  puml.generateMesh();
+}
+
+bool Mesh::checkNeighbors() const {
+  bool result = true;
+  unsigned int faceids[4]{};
+  int neighbors[4]{};
+  // check all cells
+  for (size_t elemIdx = 0; elemIdx < numElements(); elemIdx++) {
```
Comment (Thomas-Ulrich) [inline check_correctness/src/mesh.cpp]: where is this numElements defined?

File: check_correctness/src/mesh.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      if (sideBC == 0 || sideBC == 3) {
+        if (neighbors[side] < 0 && !face.isShared()) {
+          logInfo() << "Element" << cell.gid() << ", side" << side << " has a" << bcToString(sideBC) << "boundary condition, but the neighborig element doesn't exist";
+          result = false;
+        }
+      }
+      // absorbing or free surface boundaries must not have neighbor elements:
+      else if (sideBC == 1 || sideBC == 5) {
+        if (neighbors[side] >= 0 || face.isShared()) {
+          logInfo() << "Element" << cell.gid() << ", side" << side << " has a" << bcToString(sideBC) << "boundary condition, but the neighborig element is not flagged -1";
```
Comment (Thomas-Ulrich) [inline check_correctness/src/mesh.cpp]: neighboring

File: check_correctness/src/mesh.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    PUML::Neighbor::face(puml, elemIdx, neighbors);
+
+    // check all sides
+    for (size_t side = 0; side < 4; side++) {
+      const auto& face = puml.faces()[faceids[side]];
+      const auto sideBC = decodeBC(side);
+      // if a face is a regular or a dr face, it has to have a neighbor on either this rank or somewhere else:
+      if (sideBC == 0 || sideBC == 3) {
+        if (neighbors[side] < 0 && !face.isShared()) {
+          logInfo() << "Element" << cell.gid() << ", side" << side << " has a" << bcToString(sideBC) << "boundary condition, but the neighborig element doesn't exist";
```
Comment (Thomas-Ulrich) [inline check_correctness/src/mesh.cpp]: neighboring

Comment (Thomas-Ulrich) [inline check_correctness/src/mesh.cpp]: LogError if you want  that:
"The program aborts when the first face that breaks the above rules is found"?

**Review by Thomas-Ulrich — CHANGES_REQUESTED** [review]
Nice!
(I did not test it yet).
I would also integrate checkNeighbors into SeisSol (this would solve
https://github.com/SeisSol/SeisSol/issues/160 )

File: check_correctness/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+Note:
+Only regular (0), free-surface (101), dynamic-rupture (103) and absorbing (5) boundary conditions are considered.
+This does not work with special boundaries, like fault-tagging or the free-surface with gravity (yet).
```
Comment (krenzland) [inline check_correctness/README.md]: Why not? Fault-tagging is same logic as DR, boundary conditions also always have the same logic as free-surface

File: check_correctness/README.md (hunk 3 above)
Comment (sebwolf-de) [reply check_correctness/README.md]: No, it runs until the end and then exits with an error code.

File: check_correctness/src/mesh.cpp (hunk 5 above)
Comment (sebwolf-de) [reply check_correctness/src/mesh.cpp]: No, but I don't do a partitioning. Rank X reads some cells and keeps exactly these cells.

File: check_correctness/README.md (hunk 3 above)
Comment (Thomas-Ulrich) [reply check_correctness/README.md]: ok then 
"The program aborts, when one or several faces are found breaking the rules above."

File: check_mesh/README.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+```
+
+## Description
+
+1. Read `mesh.h5`
+2. For each element check that:
+  - if a face is an internal face, the neighbor *has to exist*.
+  - if a face is an external face, the neighbor *must not exist*.
+3. If an element breaks above rules, the program writes an error message.
+3. If all elements are orrect, the programm exits with a success output. If at least one element is broken, the program returns with an error code.
```
Comment (Thomas-Ulrich) [inline check_mesh/README.md]: orrect -> correct

File: check_mesh/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+Internal faces are:
+* Regular
+* Dynamic Rupture
+* Dynamic Rupture with Fault Tagging
+External Faces are:
```
Comment (Thomas-Ulrich) [inline check_mesh/README.md]: Faces -> faces

File: check_mesh/src/mesh.h [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  PUML::TETPUML puml;
+
+  enum class BCType {
+    internal, external, unknown
+  };
+
+  BCType bcToType(int id) const {
+    if (id == 0 || id == 3 || id > 64) {
+      return BCType::internal;
+    } else if (id == 1 || id == 5 || id == 6 || id == 6) {
```
Comment (Thomas-Ulrich) [inline check_mesh/src/mesh.h]: id == 6 || id == 6) ?

File: check_mesh/src/mesh.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+#pragma clang diagnostic pop
+
+Mesh::Mesh(const std::string& fileName) {
+  puml.open((fileName + ":/connect").c_str(), (fileName + ":/geometry").c_str());
+  puml.addData((fileName + ":/group").c_str(), PUML::CELL);
+  puml.addData((fileName + ":/boundary").c_str(), PUML::CELL);
+  const auto numTotalCells = puml.numTotalCells();
+  std::vector<int> cellIdsAsInFile;
+  cellIdsAsInFile.reserve(numTotalCells);
+  std::iota(cellIdsAsInFile.begin(), cellIdsAsInFile.begin(), 0);
```
Comment (Thomas-Ulrich) [inline check_mesh/src/mesh.cpp]: cellIdsAsInFile.end()?

**Review by Thomas-Ulrich — APPROVED** [review]
I found a few typos.

File: check_mesh/src/mesh.h [hunk 14]
```diff
@@ -41,6 +41,7 @@ class Mesh {
     else if (id == 3) { return std::string("dynamic rupture"); }
     else if (id == 5) { return std::string("absorbing"); }
     else if (id == 6) { return std::string("periodic"); }
+    else if (id == 6) { return std::string("free surface with gravity"); }
```
Comment (Thomas-Ulrich) [inline check_mesh/src/mesh.h:44]: id==2

File: check_mesh/src/mesh.h [hunk 15]
```diff
@@ -28,7 +28,7 @@ class Mesh {
   BCType bcToType(int id) const {
     if (id == 0 || id == 3 || id > 64) {
       return BCType::internal;
-    } else if (id == 1 || id == 5 || id == 6 || id == 6) {
+    } else if (id == 1 || id == 5 || id == 6 || id == 7) {
```
Comment (Thomas-Ulrich) [inline check_mesh/src/mesh.h:31]: 2

**Pull request merged**